### ISSUE TITLE
Compiler does not issue CS4014 despite missed await

### ISSUE THREAD

**Issue by bjornhellander** [dotnet/roslyn#45479]
**Version Used**: 
Microsoft Visual Studio Community 2019
Version 16.6.2

**Steps to Reproduce**:

1. Create a project with the following class:
```
    class MyClass
    {
        public Task<string> SomeMethodAsync()
        {
            return Task.FromResult("");
        }

        public async Task AnotherMethodAsync()
        {
            SomeMethodAsync();

            if (SomeMethodAsync() != null)
            {
            }
        }
    }

```
**Expected Behavior**:
CS4014 triggered on **both** calls to SomeMethodAsync, since the task is not awaited in either case, leading to the program continuing without waiting for any of the tasks to complete and having the same problems with any raised exceptions in both calls.

**Actual Behavior**:
CS4014 only triggered on the first call to SomeMethodAsync.

**Comment by sharwell** [comment]
Duplicate of #20782

**Comment by bjornhellander** [comment]
Ok.. The answer in #20782 seems to focus on whether or not the **called** method is defined using the async keyword. Given that logic, the second call in my example shouldn't trigger the warning either, but it does. So I'm confused.

I didn't add the async keyword on the called method in my example, because it didn't make any difference at all to the compiler's behaviour. I see exactly the same behaviour in this class, i.e. **only a warning on the first call** but not on the second:

```
    class MyClass
    {
        public async Task<string> SomeMethodAsync()
        {
            return "";
        }

        public async Task AnotherMethodAsync()
        {
            SomeMethodAsync(); // Triggers warning

            if (SomeMethodAsync() != null) // Does not trigger warning
            {
            }
        }
    }
```

The compiler **sometimes** complains about a missing await in calls from an async method to a method returning Task<T> regardless if the called method is marked as async or not.

Could you please have another look at this @sharwell, to see if you still feel the same way after adding the updated code example?

**Comment by sharwell** [comment]
It could be different so I'll let **Area-Compilers** check it out.

**Comment by jaredpar** [comment]
The compiler logic here is to trigger a warning when an `async` method is invoked as a statement expression. In this particular case the return value of the method is implicitly ignored and the method declaration declared `async` hence the compiler chooses to warn users they may have introduced a bug. 

The other cases mentioned here are not this because they're invoking the method in a context where the return value is considered. Hence the warning doesn't trigger as designed. 

Users can even avoid this warning by **explicitly** discarding the value. 

```c#
SomeAsyncMethod(); // warning
_ = SomeAsyncMethod(); // this is fine
```

As mentioned in #20782 this logic is not going to be changed at this point because it would be a major breaking change.
